### ISSUE TITLE
Mac也是按这个步骤来吗，改成CPU训练就可以了嘛

### ISSUE THREAD

**Issue by Acheng97** [sml2h3/dddd_trainer#37]

**Comment by learnprogram2** [comment]
是的，但是时间很慢

**Comment by CHINAYNINI** [comment]
速度还可以吧MBP M2P芯片，12000个样本，十几分钟就好了

**Comment by learnprogram2** [comment]
> 速度还可以吧MBP M2P芯片，12000个样本，十几分钟就好了

@CHINAYNINI 
请问可以看一下我这个为什么不可以么？将近20个小时了，然后log：
```text
2023-08-25 14:31:56.929 | INFO     | utils.train:start:137 - [2023-08-25-14_31_55]      Epoch: 84       Step: 78000     LastLoss: 0.00012210002751089633      AvgLoss: 0.0001903036296425853  Lr: 0.004640778883287701        Acc: 0.90625
2023-08-25 14:33:57.223 | INFO     | utils.train:start:108 - [2023-08-25-14_33_55]      Epoch: 84       Step: 78100     LastLoss: 0.0002089805930154398       AvgLoss: 0.00016942865564487875 Lr: 0.0045479633056219465
2023-08-25 14:35:48.219 | INFO     | utils.train:start:108 - [2023-08-25-14_35_47]      Epoch: 84       Step: 78200     LastLoss: 0.00014516110240947455      AvgLoss: 0.00017116148985223844 Lr: 0.0045479633056219465
```
我的配置没变动，我以为target中的epoch是20，现在已经84了。 也是用的测试数据集
配置TARGET: {Accuracy: 0.97, Cost: 0.05, Epoch: 20}

**Comment by enphit** [comment]
我这里100个图片，32G I9  Mac跑了三天还没完又是什么鬼？
2023-09-18 09:30:04.030 | INFO     | utils.train:start:108 - [2023-09-18-09_30_03]      Epoch: 141633   Step: 424900    LastLoss: 1.9247649106546305e-05        AvgLoss: 1.7899745298564085e-05 Lr: 0.0001380155583766437

**Comment by MaggieLOL** [comment]
在colab上跑acc正常，把整个项目原封不动复制到本地mps跑acc就一直0.0...

**Comment by kvii** [comment]
现代的模型已经用 cpu 训练的速度已经没法接受了。

其实 pytorch 是支持 mac 的，[官网链接](https://pytorch.org/get-started/locally/#start-locally)里都有 mac 版的下载方式

mac 下不用 cuda，只要声明 device 为 `pytorch.device("mps")`  就可以使用 mac 的 gpu 了。我没用过这个工程训练，不过我认为改这里的源码应该就能生效了。

https://github.com/sml2h3/dddd_trainer/blob/745c29e9a3c1e82c874d6a0f5306dda5121984eb/nets/__init__.py#L191-L196

**Comment by tzbo** [comment]
> 现代的模型已经用 cpu 训练的速度已经没法接受了。
> 
> 其实 pytorch 是支持 mac 的，[官网链接](https://pytorch.org/get-started/locally/#start-locally)里都有 mac 版的下载方式
> 
> mac 下不用 cuda，只要声明 device 为 `pytorch.device("mps")` 就可以使用 mac 的 gpu 了。我没用过这个工程训练，不过我认为改这里的源码应该就能生效了。
> 
> https://github.com/sml2h3/dddd_trainer/blob/745c29e9a3c1e82c874d6a0f5306dda5121984eb/nets/__init__.py#L191-L196

You are right. Just change to 
``` python
    def get_device(gpu_id):
        import platform
        if platform.system() == 'Darwin':
            device = torch.device("mps")
        elif gpu_id == -1:
            device = torch.device('cpu'.format(str(gpu_id)))
        else:
            device = torch.device('cuda:{}'.format(str(gpu_id)))
        return device
```

**Comment by grassmu** [comment]
在本地搞了几张图测试了下，按照步骤在设置cache时就报错了
2024-07-06 23:06:57.984 | INFO     | __main__:__init__:12 -
Hello baby~
2024-07-06 23:06:57.985 | INFO     | __main__:cache:20 -
Caching Data ----> vip_site
Path ----> ./images_set/
2024-07-06 23:06:57.986 | INFO     | utils.cache_data:__get_label_from_name:36 -
Files number is 2.
100%|██████████████████████████████████████████| 2/2 [00:00<00:00, 40721.40it/s]
2024-07-06 23:06:58.001 | INFO     | utils.cache_data:__collect_data:92 -
Coolect labels is [" ", "蹬", "块", "抢", "朽", "高", "搪", "撒", "屿", "傍", "沫"]
2024-07-06 23:06:58.002 | INFO     | utils.cache_data:__collect_data:96 -
Writing Cache Data!
2024-07-06 23:06:58.003 | INFO     | utils.cache_data:__collect_data:98 -
Cache Data Number is 2
2024-07-06 23:06:58.003 | INFO     | utils.cache_data:__collect_data:99 -
Writing Train and Val File.
2024-07-06 23:06:58.003 | ERROR    | utils.cache_data:__collect_data:106 - val setting vaild!

最后报错ERROR   val setting valid

请问这种如何解决

**Comment by kmkm209** [comment]
> 现代的模型已经用cpu训练的速度已经无法接受了。
> 
> 其实pytorch是支持mac的，[官网链接](https://pytorch.org/get-started/locally/#start-locally)里有mac版本的下载方式
> 
> mac下不用cuda，只要声明设备为`pytorch.device("mps")` 就可以使用mac的gpu了。我没用过这个工程训练，不过我认为改一下这里的源码应该就能生效了。
> 
> [dddd_trainer/nets/__init__.py](https://github.com/sml2h3/dddd_trainer/blob/745c29e9a3c1e82c874d6a0f5306dda5121984eb/nets/__init__.py#L191-L196)
> 
> 第 191 至 196 行 [745c29e](/sml2h3/dddd_trainer/commit/745c29e9a3c1e82c874d6a0f5306dda5121984eb)
> 
>  def  get_device（gpu_id）：
>      如果 gpu_id  ==  - 1：
>          设备 =  torch .设备（'cpu' .格式（str（gpu_id）））
>      别的：
>          设备 =  torch.device ( 'cuda:{}' . format ( str ( gpu_id ) ) )
>      返回 装置

感谢你的方法，很有用

<img width="999" alt="Image" src="https://github.com/user-attachments/assets/525998c3-dc8a-41f4-80f3-0ac8e8534f20" />
